qiskit/providers/options.py

```diff
@@ -26,10 +26,122 @@ class Options(types.SimpleNamespace):
+
+    def __repr__(self):
+        items = (f"{k}={v!r}" for k, v in self._fields.items())
+        return "{}({})".format(type(self).__name__, ", ".join(items))
+
+    def __eq__(self, other):
+        if isinstance(self, Options) and isinstance(other, Options):
+            return self._fields == other._fields
+        return NotImplemented
+
+    def set_validator(self, field, validator_value):
+        """Set an optional validator for a field in the options
+
+        Setting a validator enables
+        If a validator is already present for the specified field it will
+        be silently overriden. For example if you have a numeric field like
+        ``shots`` you can specify a bounds tuple that set an upper and lower
+        bound on the value such as::
+
+            options.set_validator("shots", (1, 4096))
+
+        In this case whenever the ``"shots"`` option is updated by the user
+        it will enforce that the value is >=1 and <=4096. A ``ValueError`` will
+        be raised if it's outside those bounds.
+
+        Args:
+            field (str): The field name to set the validator on
+            validator_value (list or tuple or type): The value to use for the
+                validator depending on the type indicates on how the value for
+                a field is enforced. If a tuple is passed in it must have a
+                length of two and will enforce the min and max value
+                (inclusive) for an integer or float value option. If it's a
+                list it will list the valid values for a field. If it's a
+                ``type`` the validator will just enforce the value is of a
+                certain type.
+        Raises:
+            KeyError: If field is not present in the options object
+            ValueError: If the ``validator_value`` has an invalid value for a
+                given type
+            TypeError: If ``validator_value`` is not a valid type
+        """
+
+        if field not in self._fields:
+            raise KeyError("Field '%s' is not present in this options object" % field)
+        if isinstance(validator_value, tuple):
+            if len(validator_value) != 2:
+                raise ValueError(
+                    "A tuple validator must be of the form '(lower, upper)' "
+                    "where lower and upper are the lower and upper bounds "
+                    "inclusive of the numeric value"
+                )
+        elif isinstance(validator_value, list):
+            if len(validator_value) == 0:
+                raise ValueError("A list validator must have at least one entry")
+        elif isinstance(validator_value, type):
+            pass
+        else:
+            raise TypeError(
+                f"{type(validator_value)} is not a valid validator type, it "
+                "must be a tuple, list, or class/type"
+            )
+        self.validator[field] = validator_value
+
     def update_options(self, **fields):
```

Comment:
Should this just be changed to just `update` like for a dict?

Reply:
We can add an `update()` method now, but this is part of the API and would require a deprecation to remove it. I'm not sure it's worth the cost to rename though.